pkg/binary/envoy/debug/pii.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2019 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug
+
+import (
+	"crypto/sha256"
+	"fmt"
+
+	"bitbucket.org/creachadair/shell"
+)
+
+// defaultHash returns the hashed value of s using sha256 defaultHash function
+// TODO: salt the Hash
+//nolint: unused,deadcode
+func defaultHash(s string) string {
+	h := sha256.New()
+	_, _ = h.Write([]byte(s))
+	return string(h.Sum(nil))
+}
+
+var defaultFormat = `[%START_TIME%] "%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%"` +
+	` %RESPONSE_CODE% %RESPONSE_FLAGS% "%DYNAMIC_METADATA(istio.mixer:status)%" "%REQ(USER-AGENT)%"`
+var defaultPII = map[string]bool{
+	"[%START_TIME%]": true,
+}
+
+// Filter filters log fields using pii and modify all PII fields using hash
+type Filter struct {
+	hash   func(string) string
+	pii    map[string]bool
+	format []string
+}
+
+// NewFilter constructs a custom filter object
+func NewFilter(formatStr string, hash func(string) string, pii map[string]bool) (Filter, error) {
+	// splitting formats and handle error
+	format, ok := shell.Split(formatStr)
+	if !ok {
+		return Filter{}, fmt.Errorf("error in splitting format string: %s", format)
+	}
+	return Filter{hash: hash, pii: pii, format: format}, nil
+}
+
+// process logs with hash the filter, assumes that filter has valid fields
+func (filter Filter) process(logs []string) []string {
+	out := make([]string, 0, len(logs))
+	for _, log := range logs {
+		fieldValues, ok := shell.Split(log)
+		if !ok {
+			fmt.Printf("error in splitting log: %s", log)
```

Comment:
Ah, good time to bring up two other points:
1) If we're handling PII, we shouldn't print it to the terminal. So in this case we should just print we're skipping as `infof` or if we're in debug mode the full log `debugf`:
```go
fieldValues, ok := shell.Split(log)
if !ok {
	if log.DebugEnabled() {
		log.Debuf("error splitting log, skipping: %s", log)
	} else {
		log.Info("error splitting log, skipping")
	}
	continue
}
```

2) When we're handling strings to print them for consumption and we're using `fmt.Printf`, we should prefer `%q` to `%s`. `%q` prints the bytes passed as a string, printing them in quotes and escaping quotes in the string. `%s` just prints the bytes as-is. See the differences here: https://play.golang.org/p/xtbVjA6zA3n . Since this is logging, however, we should stick with `%s`. For example, think how frustrating it would be to try and debug the tool if it printed out `error splitting log: "asdf \"asdf\""` and you put that string into the tool again and it worked, because of the escapes `%q` added, but breaks with the _real_ string: `"asdf "asdf""`. For debug data, we want closer to the golang structs we're reasoning about when we write the code - you could even make the strong argument for using `%v` in this logging context.